### ISSUE TITLE
First English Voice Library

### ISSUE THREAD

**Issue by Patchethium** [VOICEVOX/voicevox#542]
## Contents

@Hiroshiba 

According to #498 , it might be the time to take English into consideration.

I suggest we start with the corpus [LJSpeech](https://keithito.com/LJ-Speech-Dataset/), since it's in public domain we can literally do anything with it. As for the frontend, [espenet](https://github.com/espnet/espnet) recommends [g2p](https://github.com/Kyubyong/g2p), you may want to take a look at it.

I can help with the UI and Engine part if you need any, but since the recipe needed to train the core library is private, your effort is also necessary if we want to produce a functional out-of-box English TTS.

TODOs:
- [ ] UI
  - [ ] i18n / translation
    - [ ] interface translation
    - [ ] change primary key of hotkeys to English
    - [ ] translation of help information
  - [ ] new accent panel for English
  - [ ] language information in speaker meta data
- [ ] Engine
  - [ ] integrate g2p as English frontend
  - [ ] inference with English library
- [ ] Core
  - [ ] train a model with LJSpeech
- [ ] Other
  - [ ] decide on the character

This debut may hopefully raise some discussions in English community and ease new libraries to come by means like crowd funding or donations.

It's a long way to go, feel free to modify the list or bring up any suggestions.

### Pros

Anything you can expect from an English library

### Cons

We need to write the code :(

**Comment by Hiroshiba** [comment]
issueありがとうございます！
僕はちょっと日本語版TTSで手一杯ですが、誰かやりたい方が居たら、サポートを考えたいです。
ということでこのissueはcloseせず、時か人が来るまでopenし続けようと思います。

> new accent panel for English

UIの中でこれが最もアイデア力が試されると思います。
英語の知識があまりないのですが、おそらく日本語と違って英語は１単語の中に２箇所以上のアクセントがありうるので、納得のあるUIが必要そうです。
あと抑揚を音高ではなく強弱で表現するので、そこもどうすべきか考える必要があるかもしれません。

>  train a model with LJSpeech

たぶんLJSpeechであれば学習済みモデルが配布されているはずなので、trainはしなくてもとりあえず作成可能かもしれません。

---

Thank you for the issue!
I've got my hands full with the Japanese TTS, but if anyone wants to try it, I'd like to consider supporting it.
I'm not going to close this issue, but I'll keep it open until someone wants to do it.

> new accent panel for English

I think this is the most challenging UI to come up with.
I don't have a lot of knowledge of English, but unlike Japanese, there can be more than one accent in a word in English, so I think we need a UI that makes sense.
Also, intonation is represented not by pitch but by intensity, so we may need to think about how to do that as well.

> train a model with LJSpeech

Perhaps a trained model with LJSpeech should be distributed, so it may be possible to create one without training.

**Comment by Patchethium** [comment]
> たぶんLJSpeechであれば学習済みモデルが配布されているはずなので、trainはしなくてもとりあえず作成可能かもしれません。

I don't think so... Not all of the TTS systems are focusing on allowing users to adjust the parameters themselves. Take an example of Tacotron, completely E2E, that is, I could grab a Tacotron LJSpeech pretrained model and glue it to this GUI, but users will never take advantage of adjusting parameters themselves, which makes it meaningless. Actually, VOICEVOX is the only open source repository as far as I know to have such space big enough to make a difference. So if we are to make it, we should make a VOICEVOX model. (or whatever you call the models in voicevox_core)

> あと抑揚を音高ではなく強弱で表現するので、そこもどうすべきか考える必要があるかもしれません。

Although Stress replaces the Accent in Japanese, the rest of them can just stay the same. Since the stress information is already provided by g2p, I suppose it won't require much work to move the training code to English.

The interface design is pretty easy since it's just numbers representing different stress state, I may start working on it after I finished i18n.

> 僕はちょっと日本語版TTSで手一杯ですが、誰かやりたい方が居たら、サポートを考えたいです。

The problem here is the English speaking developers we need to produce an English library could only be gathered by producing one 😅 

Anyway, it'll be slow but I CAN finish these stuffs by myself, the only thing I can't achieve is to produce a VOICEVOX model with LJSpeech, where exactly I'm asking for your help. Please consider taking a look into the subject when you have the time, meanwhile I'll be working around with other stuffs on the TODO list.

**Comment by Hiroshiba** [comment]
> I don't think so... Not all of the TTS systems are focusing on allowing users to adjust the parameters themselves.

たしかに！！
仰るとおりで、VOICEVOXのようにユーザーが細かく調整できるようにするためには、よくあるE2E手法はダメなので、専用にtrainする必要があります。

You're right!
In order to allow users to make fine tuning adjustments like VOICEVOX, the common E2E method is not possible, so it needs to be specially train.

あなたの熱意を受け取りました。
このissueはこのままopenし続けます。
英語TTSのモチベーションがある機械学習エンジニアが現れることを願っています。
I received your enthusiasm.
I will keep this issue open.
I hope there will be some machine learning engineers who are motivated by English TTS.

**Comment by Patchethium** [comment]
So are you implying that a ml engineer can already reproduce the training? I though the training recipe was private at your place...

**Comment by Hiroshiba** [comment]
私のコードは日本語に特化しているので、英語では全く太刀打ちできないと思います。
なのでたぶん、誰がやるにせよフルスクラッチになりそうです･･･。

---

My code is Japanese-specific, and I don't think my code can compete at all in English.
So it's probably going to be full scratch, whoever does it...

**Comment by Patchethium** [comment]
I doubt that would take much work to adapt to new languages... TTS systems with deep learning seem to be mostly just vary from frontends when it comes to language difference. In this case, replacing the accent to stress information should be just enough. Maybe you can take a little deeper look at the topic?
Or if I have to build it from scratch, could you share some information about how the training works?

**Comment by Hiroshiba** [comment]
いえいえ！

機械学習で高品質なTTSを実現しようとするのは、本当に大変です。
そもそも機械学習はデータ作成、コードの実装、訓練、チューニング、アプリケーション化の５段階くらいがあります。

英語対応で課題になるのはデータ作成とチューニングです。
まずどんなデータ構造を入力するのが妥当なのかよく考えます。ここで間違えるとできることもできなくなります。
チューニングではどこが悪いのかを理屈で考え、それまでのコードやデータを見直す必要があります。
このサイクルを何度もやってやっと完成です。

僕は日本語TTSを作るのに半年かかりました。
ソースコードなどが揃っているので初期コストは下げられますが、たかだか１ヶ月くらいだと思います。 👍 

---

No, no, no!

Trying to achieve high quality TTS with machine learning is a real challenge.
To begin with, machine learning has about five stages: data creation, code implementation, training, tuning, and application.
If we were to adapt the VOICEVOX training model to English, the challenges would be data creation and tuning.

The first step is to think carefully about what kind of data structure is appropriate to input. If you make a mistake here, you will not be able to do what you can do.
For tuning, you need to think about what is wrong with the logic and review the previous code and data.
After doing this cycle many times, you are finally finished.

It took me half a year to make a Japanese TTS.
The initial cost can be lowered because the source code and other information is already available, but I think it can only be reduced to about 2 weeks. 👍

**Comment by Patchethium** [comment]
なるほどです、詳細な解説ありがとうございます。

こうなると、もし有志な英語TTS開発者さんが来ればいいのですが、まず自分でやってみることにします。

そして、その開発者さんのためにこのissueをopenのままにしたいんですが、どうでしょう。

**Comment by Hiroshiba** [comment]
承知しました！　このissueはopenにし続けたいと思います。

**Comment by KajMarfil** [comment]
Hello! I'm no TTS developer, but I have an idea that might open VoiceVox to any language in the world.

I am an active user of DeepFaceLab (DFL), a machine learning program that replaces one person's face with another. This is also known as "deepfaking". The program is structured in such a way that anyone can create and train a model without prior programming knowledge. All the user has to do is give the AI the source and target material. Once these two pieces of data are provided, the user can prepare the data for training by extracting, enhancing, sorting, and masking. At the start of training, the user can configure a variety of settings, from the type of architecture to the control of the Generative Adversarial Network (GAN).

I believe VoiceVox can incorporate some of these techniques used in DFL. Instead of having developers program and train their own models, they would design a separate program that would allow users to easily create models and have more fine-tuned control. The user can choose any sound source data and any language data they want. Models created in this separate program can be easily plugged into VoiceVox for use by other users.

The VoiceVox community, and other TTS communities like it, will then be split into two categories. You'd have people who create models and people who use the models. The users of VoiceVox don't need to learn how to create models if they don't want to. They can simply download and use community made models. This is also how the 3D animation/modeling community works. Some people create 3D model assets and others use these 3D model assets to create animations.

In conclusion, I believe VoiceVox would benefit greatly by allowing people to diversify into specific professions if they so please.

Sorry for the long explanation. I just thought this would be a good idea to implement. Thank you for reading! 🙂

**Comment by Patchethium** [comment]
Thanks for sharing your experience at DFL, such an active community sounds pretty promising.

I was developing my own model since the last time I left a comment under this issue. I'm also planning to publish my training recipe, hopefully such a community will then grow around it.

**Comment by Patchethium** [comment]
Good to know you're picking up English. Unfortunately I may not be able to provide much help, since I'm currently still working on my tts.

Anyway, for [1] the phonemizer, I recommend [g2p_en](https://github.com/Kyubyong/g2p), for [2] the force aligner, I recommend [mfa](https://montreal-forced-aligner.readthedocs.io/en/latest/).

`g2p_en` uses `cmudict` phoneme set, in which stress is a phoneme wise identicator, for example, in `IH0` and `AH2`, `IH` and `AH` is the phoneme and `0` and `2` is stress. As you can see, in `cmudict` it's very easy to encode stress into a one-hot and cat. it to phoneme vec. But you may also want to attach the word start and end.

It's written in numpy, shouldn't be a problem to embed it in voicevox engine.

You may want to read the doc before using `mfa`, you can use mfa's `ARPA` model, which is basically just `cmudict`. Luckily they has [this guide](https://montreal-forced-aligner.readthedocs.io/en/latest/first_steps/index.html#first-steps-align-pretrained). BTW I myself was using an unsupervised alignment search introduced in arxiv:2108.10447.

English TTS apps also suffer from the disambiguation problem in g2p (i.e the (/ri:d/) and (/rɛd/) stuff) and have different approaches. In free apps like google and microsoft azure tts free version they ignore the problem. [coqui-studio](https://coqui.ai/) and [sonantic](https://app.sonantic.io/) allow user to modify the phoneme, even pitch, energy and duration. [Elevenlab](https://beta.elevenlabs.io/) has the best approach IMO, in which they somehow combines nlp with tts and elimates the g2p stage.

**Comment by Hiroshiba** [comment]
@Patchethium 

Thank you so much for your detailed explanation!
I completely forgot about mfa. I didn't know the documentation was so comprehensive; it's really amazing.

I've tried out several TTS services, and they've been very helpful!
While `coqui-studio` doesn't allow for phoneme modification, it does enable adjustments to pitch and energy for individual phonemes, which seems to make it possible to change stress placement. The UI is similar to CeVIO's.
`Elevenlab` is an end-to-end system, and while it doesn't allow user adjustments, the quality feels quite high.

I forgot to ask one crucial question.
English-speaking regions seem to have highly developed end-to-end speech synthesis, but unlike Japanese Hiragana, English text doesn't have a one-to-one correspondence with pronunciation. Therefore, I think it wouldn't be easy to fix mispronunciations. Do people who actually use end-to-end TTS not find this inconvenient?

---

<details>
<summary>日本語</summary>

丁寧に教えて頂きありがとうございます！！
`mfa`を忘れていました。ドキュメントが充実しているのは知りませんでした、とても素晴らしいです。

いくつかのTTSサービスを試しに使ってみました。とても参考になります！
`coqui-studio`は音素の修正はできませんでしたが、音素ごとのピッチ・エナジー調整ができるのでストレス位置の変更はできそうでした。UIはCeVIOのものが近いですね。
`Elevenlab`はe2eで、ユーザー調整はできませんでしたが品質は高く感じました。

もう１つ肝心なことを質問していませんでした。
英語圏はe2e音声合成がかなり発達していそうですが、日本語と違ってテキスト（ひらがな）と読みが一対一対応していませんよね。なので読み間違いが発生したときに修正が簡単ではないと思います。
実際にe2e TTSを使う人は不便に感じていないのでしょうか･･･？

</details>

**Comment by Patchethium** [comment]
> Therefore, I think it wouldn't be easy to fix mispronunciations

No, not really, kids learn words with the help of pronouncation symbols at school, like /ri:d/. We don't usually use them in daily life like hiragana in japanese, but we know how to use when necessary.

You're right about that almost every English TTS app doesn't include a phoneme editing feature, but I don't think they leave the problem there just because they don't have other ways. One reason is that some of them need to sell this feature to the enterprise, like `sonantics`, another reason is that it's just too trivial to do so. What they focus on is speech synthesis, end user accessbility has to step aside.

For example, I do feel frustrated when Eleven labs misspells a phone number as demical, but it can be fixed very easily by typing a space between each characters. Fixing spelling is not a big issue, they can add it anytime with a few lines of code, but their **cross lingual zero-shot voice cloning** and **exetremely natural voice** are certainly huge issues.

PS: Just a thought, I'm also start thinking the pitch/duration tuning in the same way, when the vocie is not natural enough, we tune the parameters. What if we can already achieve a naturalness higher than human tuning? No offense, while tuning `VOICEVOX`, most of my time is spent on making the voice more **natural**, not **better** or **more suitable to the scene**. And the final result still feels lower than `Elevenlab`'s. I'd say tuning is a sweet girl, but should we put her at the center of the stage? I don't think so.

**Comment by Hiroshiba** [comment]
Thank you for the detailed explanation! It has definitely helped me get a clearer image of English TTS!

My thoughts might be a bit outdated, but I do have some unique ideas.

What I want to create is a "character voice synthesis" system (similar to VOICEROID), with the main use case being video game commentary. In a domain with many unknown words like this, I believe it's difficult to synthesize natural-sounding speech without user adjustments. For example, in a Pokémon game commentary, there should be a lot of unknown words. I believe that an application that can correct the pronunciation of unknown words is essential for it to be on the stage.

Additionally, I want to make it usable on regular people's local computers, which means that large models can't be used. Achieving both a small model size and naturalness at the same time is challenging, so I'm considering sacrificing some default voice quality for a smaller model size. This might not be much of an issue if there's a large amount of data available.

Well, having different opinions isn't necessarily a bad thing. We can simply adjust the UI according to the voice synthesis model...!

**Comment by Patchethium** [comment]
I'm not saying tuning isn't sweet, but just explaining why the big enterprises don't have interest in improving user accessibility. I'd definitely be happy to see an English TTS with pitch tuning, allows to edit phoneme, runs locally so I don't have to tolerate `coqui-studio`'s 502 bad gateway.

BTW saying Pokemon, I copied a few pokemon's names into Elevenlab and it says them all very properly. I think it's because g2p in English itself is an easy task, I once made a single LSTM seq2seq model to do train on cmudict, and it already got 90% accuracy on test set.

**Comment by Hiroshiba** [comment]
I see... Are those recent Pokemon names that are not included in the dictionary? If so, it might be that there's a very low chance of getting them wrong in the first place.

<details>
<summary>日本語</summary>

なるほど･･･。それは（辞書に含まれていない）最近のポケモンの名称でしょうか。だとしたらそもそも間違うこと自体がとても少ないのかもしれませんね･･･。

</details>

**Comment by Patchethium** [comment]
`Elevenlabs` in deed has a good g2p model, it can even handle nonsense words I ask chatgpt to make up like `Flimble` and `Quink`, very decently. G2p in English is not a problem, as I mentioned.

**Comment by Hiroshiba** [comment]
That's impressive. Do you happen to know if `g2p_en` achieves similar accuracy? If it performs at the same level as `Elevenlabs`, it seems like we could lower the priority of the reading correction feature a bit.

<details>
<summary>日本語</summary>

素晴らしいです。`g2p_en`も同じくらいの精度が出るかご存知だったりしますか？
`Elevenlabs`と同じくらいの性能なのであれば、読みの修正機能は少し優先度を下げても良さそうに思いました。

</details>

**Comment by Patchethium** [comment]
I don't think we need to spend more time on this topic. As I've said before, reading correction is always better than nothing and I am 100% supporting fitting one into the GUI. Just set off and go already, contributors will add these if they think necessary.

**Comment by hayabusa255** [comment]
現在個人的にこれに取り組んでいるものです。
@Hiroshiba さんの配信にたまにお邪魔させていただいてもいます

とりあえず、日本語版VoiceVoxをもとに英語版の構造を考えてみました。
![image](https://github.com/user-attachments/assets/d28a03f0-d0a9-4e49-bdff-1e973463d580)

現在音素から、継続長を取り出すモデルを作成及び評価中です。

特徴量からメルスペクトログラムを予想するモデルを作るところが今後課題となりそうなのでそこまで進んだら色々お聞きしたいです。

自分自身もmlは初めて触るので、学びながら高品質のモデルを作成できるよう頑張ります。

目標としてはLJSpeechをベースに作成して、既存の日本語キャラの音声データを用いて転移学習を行い、ずんだもんや、四国めたんが英語を流暢に話せるモデルを作成して実装することです。

@Patchethium さんは英語版TTSに詳しそうなので、もしよければいっしょに共同開発を行いたいと考えています。

```English

I am currently working on this personally.
I occasionally join Hiroshiba-san's streams.

For now, I've come up with a structure for the English version of VoiceVox based on the Japanese version.
![image](https://github.com/user-attachments/assets/d28a03f0-d0a9-4e49-bdff-1e973463d580)

I am currently creating and evaluating a model that extracts phoneme durations.

The next challenge will be to create a model that predicts Mel-spectrograms from features, so once I progress to that stage, I would like to ask for some advice.

This is my first time working with machine learning, but I am learning as I go, aiming to create a high-quality model.

My goal is to create a model based on LJSpeech, then perform transfer learning using the existing Japanese character voice data to make Zundamon and Shikoku Metan speak fluent English.

@Patchethium seems knowledgeable about English TTS, so if you're interested, I'd love to work together on this project.

```

**Comment by Patchethium** [comment]
I'm glad that there are other people interested in the topic and are willing to get involved. Although I have different opinions on how a TTS with characters should be (it should be guarded behind a payment wall), I am also happy to help.

The structure you show here, being a classical fastspeech2 model, is decent enough to produce a sane speech, although may not be the best. To achieve your goal (making zundamon and metan able to speak English), there are some suggestions:

Necessary
 - use domain adversarial training to close the gap between Japanese and English phonemes, see (arxiv:19), it is crucial to multilingual TTS
- finetune the hifiGAN on spectrogram produced by the acoustic model
- you didn't depict the phoneme hidden state, so I assume you know that, but in case, you need to add the pitch to it and upsample it according to the duration 
- use a bigger dataset, like Libri TTS or VCTK, LJS won't work well in a multilingual scenario

Optional (if you are familiar)
- use gaussian up-sampling for acoustic model
- change hifiGAN to vocos, which is faster and more stable in training
- quantize the pitch into a finite bin, produces higher robustness

Advanced (if you are an expert)
- use flow matching decoder, MSE is obsolete 
- use discrete audio codec instead of Mel spectrogram (what I am struggling with)
- use a language model (like bert) to improve prosody 

As I am (surprisingly) still struggling with my own TTS application, I can't really offer some help. But you can ask me anything by any means (discord or GitHub), and as long as I have time I will try to answer them.

**Comment by Hiroshiba** [comment]
英語は音素区切りやモーラ区切りではなくシラブル区切りになっていてほしい、でもシラブル区切りにできるライブラリがわからなかったので調べてみました！
せっかくなのでまとめてみます！

* [Festival](https://github.com/festvox/festival/)
	* CevIO AIが使ってるっぽい？
	* とりあえず動かす分には[phonimizer](https://github.com/bootphon/phonemizer/)が便利そう
		* 強勢（ストレス）を得るにはコードの変更が必要、Festivalは普通に出力しているのでパースすればいい
	* C言語なのでpythonラップも原理上はおそらく可能
* g2p_en
	* 強弱2つの強勢が得られるが、シラブルを得られない

何とかして2つとも使えると便利そうだけど･･･。

Festivalを使ってアクセント情報を得るメモ：
<details>

aptでfestivalをインストールしたあと、↓を`extract_accent.txt`などに保存
```
(define (phonemize line)
  (set! utterance (eval (list 'Utterance 'Text line)))
  (utt.synth utterance)
  (print (utt.relation_tree utterance "SylStructure")))

(set! lines (load "input.txt" t))
(mapcar (lambda (line) (phonemize line)) lines)
```

続いて英語テキストを`input.txt`に保存。

あとは`festival -b extract_accent.txt`を実行すればテキスト系列が得られる。
たぶん`stress 1`になってるものが強勢。

</details>

<details><summary>ChatGPT English</summary>

---

I wanted the English to be divided by syllables, not by phonemes or morae, but I didn't know which library could do that, so I researched it!
I'll summarize what I found!

* [Festival](https://github.com/festvox/festival/)
  * Seems like CevIO AI uses it?
  * For basic operation, [phonimizer](https://github.com/bootphon/phonemizer/) seems convenient
    * Modifying the code is necessary to get stress (emphasis), but Festival outputs it normally, so parsing should work
  * Since it's in C, wrapping it in Python should theoretically be possible
* g2p_en
  * Provides two levels of stress but does not give syllables

It would be useful if I could somehow use both.

Notes on using Festival to get accent information:

After installing Festival with apt, save the following in a file like `extract_accent.txt`:
```
(define (phonemize line)
  (set! utterance (eval (list 'Utterance 'Text line)))
  (utt.synth utterance)
  (print (utt.relation_tree utterance "SylStructure")))

(set! lines (load "input.txt" t))
(mapcar (lambda (line) (phonemize line)) lines)
```

Then save the English text in `input.txt`.

After that, execute `festival -b extract_accent.txt` to obtain the text sequence.
Probably, the ones marked `stress 1` are stressed.

</details>

**Comment by hayabusa255** [comment]
継続長求めるモデルもシラブル区切りでできた方が便利ですかね？
すでに音素ごとのやつはできているんですが...

**Comment by Patchethium** [comment]
No, most neural TTS papers use phoneme wise duration. In fact, I've never came across any TTS that involves syllable analysis.

**Comment by Hiroshiba** [comment]
長さは音素レベルで良いと思います！

* 子音だけ伸縮したくなることがある（日本語の例だと、「sーーーoれはどうかnーa？」みたいな表現とか、あとどうしても滑舌悪くなる時とか）
* 子音ごとに長さを求めてれば、仮にシラブル単位で伸縮したい場合はルールベースのGUI機能があればなんとかなる（たぶん。）
* なんかこっちのが楽そう。

でもちょっといろんな人にも聞いてみてみます。

**Comment by hayabusa255** [comment]
了解です。
お二人ともアドバイスありがとうございます。
とりあえずシラブル区切りでの、ピッチ予測モデルをできれば11月中にデータ収集およびトレーニング評価、完成まで目指して作成します！
g2pとFestivalの連携は考えます。

いただいたアドバイスをもとに、もう一度構成図作り直すことも考えてます。


<details>
<summary>English</summary>
Thank you both for the advice. I’ll aim to create a syllable-based pitch prediction model by the end of November. This will include data collection, training, evaluation, and completion.  
I’ll also consider integrating g2p and Festival.  


Based on the advice received, I may revise the architecture diagram.

</details>
